client/go/outline/connectivity/connectivity.go

```diff
@@ -34,54 +33,49 @@ const (
-type authenticationError struct {
-	error
-}
+// Status indicates if a proxy server can handle traffic of TCP, UDP, or both.
+// It is a bitwise combination of [TCPConnected] and [UDPConnected].
+type Status int
 
-// reachabilityError is used to signal an unreachable proxy.
-type reachabilityError struct {
-	error
-}
+const (
+	// TCPConnected indicates the proxy server can handle TCP traffic.
+	TCPConnected Status = 1 << iota
+
+	// UDPConnected indicates the proxy server can handle UDP traffic.
+	UDPConnected
+)
 
 // CheckConnectivity determines whether the Shadowsocks proxy can relay TCP and UDP traffic under
-// the current network. Parallelizes the execution of TCP and UDP checks, selects the appropriate
-// error code to return accounting for transient network failures.
-// Returns an error if an unexpected error ocurrs.
-func CheckConnectivity(client *outline.Client) (neterrors.Error, error) {
+// the current network.
+// Parallelizes the execution of TCP and UDP checks, and reports the [Status].
+// There's no error if the server at least supports TCP.
+// A [platerrors.PlatformError] will be returned if the server doesn't support TCP,
+// or if any other unexpected issues occur (such as invalid credential).
+func CheckConnectivity(client *outline.Client) (Status, error) {
 	// Start asynchronous UDP support check.
-	udpChan := make(chan error)
+	udpChan := make(chan bool)
 	go func() {
 		resolverAddr := &net.UDPAddr{IP: net.ParseIP("1.1.1.1"), Port: 53}
 		udpChan <- CheckUDPConnectivityWithDNS(client, resolverAddr)
 	}()
 	// Check whether the proxy is reachable and that the client is able to authenticate to the proxy
-	tcpErr := CheckTCPConnectivityWithHTTP(client, "http://example.com")
+	tcpErr := checkTCPConnectivityWithHTTP(client, "http://example.com")
 	if tcpErr == nil {
-		udpErr := <-udpChan
-		if udpErr == nil {
-			return neterrors.NoError, nil
+		if udpSupported := <-udpChan; udpSupported {
+			return TCPConnected | UDPConnected, nil
 		}
-		return neterrors.UDPConnectivity, nil
-	}
-	var authErr *authenticationError
-	var reachabilityErr *reachabilityError
-	if errors.As(tcpErr, &authErr) {
-		return neterrors.AuthenticationFailure, nil
-	} else if errors.As(tcpErr, &reachabilityErr) {
-		return neterrors.Unreachable, nil
+		return TCPConnected, nil
 	}
-	// The error is not related to the connectivity checks.
-	return neterrors.Unexpected, tcpErr
+	return 0, tcpErr
 }
 
 // CheckUDPConnectivityWithDNS determines whether the Shadowsocks proxy represented by `client` and
 // the network support UDP traffic by issuing a DNS query though a resolver at `resolverAddr`.
-// Returns nil on success or an error on failure.
-func CheckUDPConnectivityWithDNS(client transport.PacketListener, resolverAddr net.Addr) error {
+// Returns true on success or false on failure.
+func CheckUDPConnectivityWithDNS(client transport.PacketListener, resolverAddr net.Addr) bool {
```

Comment:
At least for now we are not passing this error back to the client at all, will it be a valid scenario in the future?

Reply:
We are not passing because the wiring is lacking, not because we don't want to.
Yes, we do want to expose more details and possibly change the UI and localize messages in some cases.